Make (equivalent for) dragAndMoveBall available again

igrep opened the pull request:
~~This is still draft because I left a several things below and I want to show you before completing to reduce the risk in case of refusal.~~ **FEEDBACK WELCOME!**

- ~~More customizable debugging mode: make `controlMode` and `visualizeBlocks` separately configurable on popup.html.~~ See https://github.com/canalun/brick-break-anywhere/pull/15#discussion_r1650971230 on why I removed this.
- [x] Refactor: apply `updateBallPositionTo` to other objects if possible and it deserves.

Background
====

When I first play with brick-block-anywhere on [my website](https://the.igreque.info/), I found a bug that could be a good chance for me to learn the browser's behavior. This feature is very useful for reproducing the bug and testing after fixing it.

Changes Summary
====

- Add a function `controlBallByMouse` to move the ball by mouse so that we can easily put the ball at some place for debugging.
- Abstract the procedure to move the ball (`updateBallPositionBy` and `updateBallPositionTo`) to hide the details and for consistency.
- Merge `TestMessage` into `StartMessage` because now the non-debugging mode has to know whether the debugging mode is enabled to disable the default movement of the ball.
    - Bonus: this change also makes https://github.com/canalun/brick-break-anywhere/commit/12c7f5388700581a373decda998a4bffbf957668  unnecessary. It's simpler!

Screenshot
====

![bba-control-by-mouse](https://github.com/canalun/brick-break-anywhere/assets/227057/8b7658c3-8bad-4f67-a88f-1be1c58f4a0a)

igrep committed:
Revert "Keep debug-or-start between replays"

This reverts commit 12c7f5388700581a373decda998a4bffbf957668.

igrep committed:
Make (equivalent for) `dragAndMoveBall` available again

Background
====

When I first play with brick-block-anywhere on [my website](https://the.igreque.info/),
I found a bug that could be a good chance for me to learn the browser's
behavior. This feature is very useful for reproducing the bug and testing
after fixing it.

Changes Summary
====

- Add a function `controlBallByMouse` to move the ball by mouse so that we
  can easily put the ball at some place for debugging.
- Abstract the procedure to move the ball (`updateBallPositionBy` and
  `updateBallPositionTo`) to hide the details and for consistency.
- Merge `TestMessage` into `StartMessage` because now the non-debugging
  mode has to know whether the debugging mode is enabled to disable
  the default movement of the ball.
    - Bonus: this change also makes https://github.com/canalun/brick-break-anywhere/pull/13/commits/12c7f5388700581a373decda998a4bffbf957668
      unnecessary. It's simpler!

Next
====

- More customizable debugging mode: make `controlMode` and
  `visualizeBlocks` separately configurable on popup.html.
- Refactor: apply `updateBallPositionTo` to other objects if possible and
  it deserves.

igrep committed:
Add a TODO

canalun commented on src/game/configuration/settings.ts:
```diff
@@ -9,7 +9,19 @@ export const collisionPointOnBallClass = "bba-collision-point-on-ball"
   withScoreboard: boolean
   initialBallSpeed: "low" | "middle" | "high" | "superHigh"
-  sound: boolean
+  sound: boolean,
+  visualizeBlocks: boolean,
+  controlMode: "normal" | "mouse"
+}
```
[imo] How about a structure like this?👀 I think it might be more robust.
```suggestion
  sound: boolean
} & {
  visualizeBlocks: false,
  controlMode: "normal"
} | {
  visualizeBlocks: true,
  controlMode: "mouse"
} 
```

igrep replied:
🤔 I'm undecided about this.

I found two bugs using the `controlMode: "mouse"` feature. I actually couldn't  find one of the bugs actually because the borders drawn by `visualizeBlocks` prevent me from reproducing it. That's why I left the TODO "More customizable debugging mode: make controlMode and visualizeBlocks separately configurable on popup.html.".

But now, I know how to fix it and the change to fix it would perhaps remove the root problem of `visualizeBlocks`. So separating them could just make the code more complex.

I beg your opinion!

NOTE: I'll send another PR to fix the bug mentioned here today or tomorrow!

igrep replied:
Rethinking this problem, now I want to keep as is.

- As my need for separating the debug options has already been satisfied by (manually editing the code), we would rarely need to switch them independently. So keep popup.tsx simple by not adding new `<input>`s for the time being.
- But IMO `visualizeBlocks` and `controlMode` are essentially different. They should be configured independently (even by temporarily updating the code). Your suggestion makes it a little harder to do that.

canalun commented on src/game/configuration/settings.ts:
```diff
@@ -9,7 +9,19 @@ export const collisionPointOnBallClass = "bba-collision-point-on-ball"
+  controlMode: "normal" | "mouse"
+}
+
+export function disableScoreboardIfDebug(options: StartOptions): StartOptions {
```
[Q] Is it possible to remove this function and make the scoreboard disabeld for the debug mode within `sendMessageToIsolatedWorldOnActiveTabForTest()` in `popup.tsx`? I'd like to hear your opnion :)

igrep replied:
👍  Indeed. I'm a little worried that'd make `popup.tsx` complex, but more user-friendly!

igrep replied:
It seems I mistook your comment wrong 😅 . But https://github.com/canalun/brick-break-anywhere/pull/15/commits/2614729ca72b29736c672db23768156bb82b9fc0 would surpass your expectation!

canalun commented on src/game/animation/updateBall.ts:
```diff
@@ -221,3 +222,25 @@ export function updateBallDirectionByCollisionWithBlocks(
   return { x: currentBallDirection.x, y: currentBallDirection.y }
 }
+
+export function updateBallPositionBy(ball: Ball, velocity: Vector): void {
+  const currentBallPosition = getBallCenterPosition(ball)
+  Object.assign(ball.style, {
+    transform:
+    `translate(` +
+      `${currentBallPosition.x - ballSetting.radius + velocity.x}px, ` +
+      `${-(currentBallPosition.y - ballSetting.radius + velocity.y)}px` +
+      `)`
+  })
+}
+
+// TODO: Also available for Bar?
+export function updateBallPositionTo(ball: Ball, position: { x: number, y: number }): void {
+  Object.assign(ball.style, {
+    transform:
+    `translate(` +
+      `${position.x}px, ` +
+      `${position.y}px` +
+      `)`
+  })
+}
```
Thanks, it's nicely comprehesive!!! :)

igrep replied:
Fixed with https://github.com/canalun/brick-break-anywhere/pull/15/commits/f29abee556dcc81ffeaaf2e11d6206ee7ea9d3c4.

canalun commented:
@igrep 
Apologies for my late reply. After a busy month, now I have time to review this PR.
I like your idea! So could you please go for it? Sorry for late response.
(FYI. I've left some comments, thank you for checking them!)

canalun commented:
And I like your branch name :)

igrep commented:
> And I like your branch name :)

Thank you! 😁  To tell the truth, I couldn't make up a good name for `controlMode: "mouse"` at first. Then I decided to give a name that sounds something super because it breaks the nature of the brick-break game. "God mode" might be more suitable, but "God" could be so religious that I couldn't determine it's safe. "Ninja" would have no such a problem and everyone loves it by contrast!

And the prefix "2" means it's just the second try 😅 .

igrep committed:
Several improvements on popup.tsx

- Disable scoreboard on UI when the debug mode is enabled, then delete `disableScoreboardIfDebug`.
    - Answer to [a review comment from canalun](https://github.com/canalun/brick-break-anywhere/pull/15#discussion_r1649213171).
        - It seems I actually took the comment wrong, but this would make it easier to add independent debug options.
- Abstract the common steps in `sendMessage*` functions.
- Add the "Start" button to the bottom
    - This makes easier to start when enabling the debug mode.

igrep committed:
Abstract the function to move ball and bar

canalun commented on src/game/animation/updateObject.ts:
```diff
@@ -0,0 +1,9 @@
+export function updateObjectPositionTo(ball: HTMLDivElement, position: { x: number, y: number }): void {
```
How about changing the name of 1st arg? :)
```suggestion
export function updateObjectPositionTo(obj: HTMLDivElement, position: { x: number, y: number }): void {
```

canalun commented on src/popup.tsx:
```diff
@@ -140,12 +156,29 @@ function IndexPopup() {
+        </div>
       ) : null}
+      <br />
+      <button onClick={start}> Start! </button>
```
It seems that the start button is duplicated unnecessarily?👀

igrep replied:
As referred in https://github.com/canalun/brick-break-anywhere/commit/2614729ca72b29736c672db23768156bb82b9fc0, it's intentional.
I think now the settings menu is long enough that putting the start button both at the top and the bottom is a good choice.

canalun reviewed (commented):
sorry for the delay. I'll review the other PRs also soon...!!

igrep committed:
Rename the argument as the function name

Co-authored-by: canalun <i.am.kanaru.sato@gmail.com>

igrep commented:
NOTE: I force-pushed https://github.com/canalun/brick-break-anywhere/commit/3aad11f1f64cb45f3fb7484bde5ad944b15cce5e because https://github.com/canalun/brick-break-anywhere/commit/80770191c2d8300b39a4786fff26eb311989a470 doesn't work because only the declaration of `ball` is renamed.

canalun reviewed (approved):
thank you so much!!!!!
